### PR TITLE
Ice sheet: consistency under dimensional rescaling and rotation

### PR DESCRIPTION
**Pull request by alex-huth**
This PR guarantees consistency of the ice sheet code under rotated/flipped input and grids, and under dimensional rescaling to the +/-140 power. Bitwise identical results were achieved under rotation/rescaling using standard 2D test cases such as MISMIP+. Major changes include the following:

- Specified order of operations throughout, and modified the implementation of FEM integration,  h-point-to-node operations, etc for consistency under rotation (see MOM_ice_shelf_dynamics.F90 subroutines calc_shelf_driving_stress, CG_action, CG_action_subgrid_basal, matrix_diagonal, shelf_advance_front, and interpolate_h_to_b).
- Added an ice-shelf version of 'first_direction' and 'alterate_first_direction', allowing users to control the order in which the x- and y-direction ice-shelf advection calls are made. Required for consistency under rotation.
- Dimensional rescaling fixes throughout MOM_ice_shelf_initialize.F90, and within MOM_ice_shelf_dynamics.F90 (see subroutines initialize_ice_shelf_dyn, ice_shelf_solve_outer, and calc_shelf_taub)
- Rotation/rescaling testing revealed two additional bugs that were subsequently fixed:
(1) An index error in the conjugate gradient algorithm (subroutine CG_action)
(2) Discretization error for the east/north fluxes in subroutines ice_shelf_advect_thickness_x and ice_shelf_advect_thickness_y.

The PR also includes several simple changes for code readability and computational efficiency:

- Saved Phi, PhiC, and Phisub in the ice shelf dynamics control structure at the start of each run, rather than reallocating and recalculating these static fields each time they are used. Reshaped the Phisub array for computational efficiency over loops.
- Modified the sub-cell grounding scheme so that it is only called for partially-grounded cells that contain the grounding line (i.e. where float_cond(i,j)==1). Added float_cond to the ice dynamics structure for output.
- Simplified the option to calculate ice viscosity at 4 quadrature points per cell rather than at cell centers, by adding parameter NUMBER_OF_ICE_VISCOSITY_QUADRATURE_POINTS and reshaping CS%ice_visc accordingly.
- Style changes and removal of unused code (e.g. removed subroutine apply_boundary_values and field thickness_bdry_val)

### CODE REVIEW

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 1]
```diff
@@ -3066,7 +3072,8 @@ subroutine calc_shelf_taub(CS, ISS, G, US, u_shlf, v_shlf)
   real :: alpha !Coulomb coefficient [nondim]
   real :: Hf !"floatation thickness" for Coulomb friction [Z ~> m]
   real :: fN !Effective pressure (ice pressure - ocean pressure) for Coulomb friction [R L2 T-2 ~> Pa]
-  real :: fB !for Coulomb Friction [(L T-1)^CS%CF_PostPeak ~> (m s-1)^CS%CF_PostPeak]
+  real :: fB !for Coulomb Friction [(T L-1)^CS%CF_PostPeak ~> (s m-1)^CS%CF_PostPeak]
+  real :: fN_scale !To convert effective pressure to mks units during Coulomb friction
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:3076]: fN_scale needs to have units defined, probably as [T2 kg R-1 L-2 m-1 s-2 ~> 1] or [Pa T2 R-1 L-2 ~> 1].  Also the documented units for fN two lines above need to be changed to just [Pa] for consistency.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 2]
```diff
@@ -1732,37 +1756,39 @@ subroutine ice_shelf_advect_thickness_x(CS, G, LB, time_step, hmask, h0, h_after
...
-          h_face = CS%thickness_bdry_val(i,j)
+          h_face = CS%h_bdry_val(i,j)
         elseif (hmask(i,j) == 1) then ! There can be eastward flow through this face.
-          if ((hmask(i-1,j) == 1) .and. (hmask(i+1,j) == 1)) then
+          if ((hmask(i-1,j) == 1 .or. hmask(i-1,j) == 3) .and. &
+            (hmask(i+1,j) == 1 .or. hmask(i+1,j) == 3)) then
             slope_lim = slope_limiter(h0(i,j)-h0(i-1,j), h0(i+1,j)-h0(i,j))
             ! This is a 2nd-order centered scheme with a slope limiter.  We could try PPM here.
-            h_face = h0(i,j) - slope_lim * 0.5 * (h0(i,j)-h0(i+1,j))
+            h_face = h0(i,j) - slope_lim * (0.5 * (h0(i,j)-h0(i+1,j)))
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:1772]: Multiplying by an integer power of 2 (e.g., 0.5 here) is order invariant, so the extra parentheses here are unnecessary but benign.

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
I have carefully examined all of these changes, and they all look good to me, apart from one omitted change in the units in a comment on line 3074 (see the specific comment).

  This will be ready to go once that very minor change is made.

**Review by Hallberg-NOAA — APPROVED** [review]

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/21727 :heavy_check_mark:

**Pull request merged**